blogposts/2021/how-to-not-break-a-search-engine-or-what-i-learned-about-unglamorous-engineering.md

```diff
@@ -0,0 +1,290 @@
+derive appreciation from users, peers, or managers when the point was for no
+one to notice anything significant had changed; when there's no perceptible
+delta?
+
+## Value in the unseen and unspoken
+
+There's unglamorous engineering in the software all around us. For all
+its lack of recognition, I wish we grasped its value a bit better. I'm reminded
+of a tweet by a former colleague who researched donations for open source
+projects:
+
+<blockquote class="twitter-tweet tw-align-center" data-conversation="none"><p lang="en" dir="ltr">I can tell you from some informal interviews we did outside that paper, that people spend the money on gruntwork — the stuff that’s fun they’re more likely to do anyway, money or not.</p>&mdash; Bogdan Vasilescu (@b_vasilescu) <a href="https://twitter.com/b_vasilescu/status/1279199236094132227?ref_src=twsrc%5Etfw">July 3, 2020</a></blockquote> <script async src="https://platform.twitter.com/widgets.js" charset="utf-8"></script>
+
+This suggests that gruntwork, if not glamorous, is certainly valuable (and
+perhaps, even disproportionately so). At the same time, I wouldn't necessarily
+call unglamorous engineering thankless. Your close peers may be very
+thankful for a component rewrite or code refactor (I know I am), but no amount
+of gratitude will make unglamorous work glamorous. It simply has to get done in
+spite of whether there's a desire to tackle it, or some incentive or
+recognition. And sometimes you might be the only one who knows it. Maybe it's
+just part of writing code. At times you need to wear the hat of a hygiene
+engineer.<a href="#footnote-2"><sup>2</sup></a> A developer-janitor. A code plumber. And that
+cost can feel very personal.
+
+The lamentable part of the unglamorous engineer's work is that there's little
+account of these feats. Technical media and blogs are prone to talk about
+features or intellectual explorations. The new and shiny is naturally engaging.
+But I also want to hear about that impressive engineering feat that no one
+noticed. The one where a developer or team pulled off some tectonic shift in a
+codebase, everyone oblivious except themselves. 
+
+## New capabilities while steadily transitioning the old
+
+Sourcegraph supports more combinations of operators now. And nothing outright
+broke in order to get there. And to be clear, I did more than rewriting and
+testing parsers in that six-month time frame. There were frequent updates in
+releases that customers could enable along the way, with a safe fallback to
+disable new functionality. I also enabled new implementations on our dogfood
+instance as things matured. The core implementation probably took only two or
+three months, but crossing the point of no return and removing the fallback was
+a slow and thorough process. There's more to tweak, but when the previous code
+was finally dropped, it felt good and it felt right.
+
+## Developer tools reduce unglamorous effort
```

Comment:
It's funny, a Sourcegraph call out isn't actually something that I feel compelled to do, but more like I realized along the way that dev tools actually have this role in bringing value.  Will rework this or just outright remove it if it doesn't work.

Reply:
I just nuked it, don't see a good way to fit it in. Adding a good concrete example doubles down on "oh yeah this is hella relevant see" and could still come across as superficial maybe. I tried a footnote but it seems like a distracting thought in context. Maybe something a fun tweet could hint at, but just going to leave it out of the blog post itself since I can see how it detracts from the other points.